### PR TITLE
fix: improvements in code block

### PR DESCRIPTION
**Pull request by jpuri**
PR has small fix around code block:

1. Table menu option is disabled when code block is selected.
2. Code block menu now toggles code block.
3. `cmd-alt-0` toggles code block.

### CODE REVIEW

**Comment by jpuri** [comment]
Fixes https://github.com/tinacms/tinacms/issues/1231

**Comment by DirtyF** [comment]
Here's a quick test:
https://www.loom.com/share/5a0f459b896c48b9b49107f2fdf35dc5

1. The selection is lost when you toggle the code block
2. You can loose content if you toggle many times ⚠️ 
3. Table is inactive when code block is active 🆗

**Comment by jpuri** [comment]
@DirtyF : selection issue is fixed in the PR.

I also want to address in this PR:
1. https://github.com/tinacms/tinacms/issues/1270
2. https://github.com/tinacms/tinacms/issues/1276

**Comment by DirtyF** [comment]
Given I typed some text to create a code block

```
10 BASIC
20 BASIC
```

When I try to edit the text in the code block
Then it throws an error

https://www.loom.com/share/e36d1f2d378a40e4bbfbb037b238df44

<details>

<summary>Error</summary>

```js
×
InternalError: too much recursion
apply
node_modules/prosemirror-state/dist/index.es.js:806

  803 | 
  804 | // :: (Transaction) → EditorState
  805 | // Apply the given transaction to produce a new state.
> 806 | EditorState.prototype.apply = function apply (tr) {
  807 |   return this.applyTransaction(tr).state
  808 | };
  809 | 

dispatchTransaction
src/components/ProsemirrorEditor/utils/buildEditor.ts:54

  51 |  * the Prosemiror state.
  52 |  * @param tr
  53 |  */
> 54 | dispatchTransaction(tr) {
     | ^  55 |   const nextState: any = view.state.apply(tr as any)
  56 | 
  57 |   view.updateState(nextState as any)

dispatch
node_modules/prosemirror-view/dist/index.es.js:4805
destroy
src/plugins/CodeBlock/nodeView.ts:39

  36 |   CodeMirror = require('codemirror')
  37 | }
  38 | 
> 39 | export class CodeBlockView implements NodeView {
  40 |   cm: CodeMirrorEditor
  41 |   dom?: HTMLElement
  42 |   updating: boolean = false

dispatchTransaction
src/components/ProsemirrorEditor/utils/buildEditor.ts:54

  51 |  * the Prosemiror state.
  52 |  * @param tr
  53 |  */
> 54 | dispatchTransaction(tr) {
     | ^  55 |   const nextState: any = view.state.apply(tr as any)
  56 | 
  57 |   view.updateState(nextState as any)

dispatch
node_modules/prosemirror-view/dist/index.es.js:4805
destroy
src/plugins/CodeBlock/nodeView.ts:39

  36 |   CodeMirror = require('codemirror')
  37 | }
  38 | 
> 39 | export class CodeBlockView implements NodeView {
  40 |   cm: CodeMirrorEditor
  41 |   dom?: HTMLElement
  42 |   updating: boolean = false

dispatchTransaction
src/components/ProsemirrorEditor/utils/buildEditor.ts:54

  51 |  * the Prosemiror state.
  52 |  * @param tr
  53 |  */
> 54 | dispatchTransaction(tr) {
     | ^  55 |   const nextState: any = view.state.apply(tr as any)
  56 | 
  57 |   view.updateState(nextState as any)

dispatch
node_modules/prosemirror-view/dist/index.es.js:4805
destroy
src/plugins/CodeBlock/nodeView.ts:39

  36 |   CodeMirror = require('codemirror')
  37 | }
  38 | 
> 39 | export class CodeBlockView implements NodeView {
  40 |   cm: CodeMirrorEditor
  41 |   dom?: HTMLElement
  42 |   updating: boolean = false

dispatchTransaction
src/components/ProsemirrorEditor/utils/buildEditor.ts:54

  51 |  * the Prosemiror state.
  52 |  * @param tr
  53 |  */
> 54 | dispatchTransaction(tr) {
     | ^  55 |   const nextState: any = view.state.apply(tr as any)
  56 | 
  57 |   view.updateState(nextState as any)

dispatch
node_modules/prosemirror-view/dist/index.es.js:4805
destroy
src/plugins/CodeBlock/nodeView.ts:39

  36 |   CodeMirror = require('codemirror')
  37 | }
  38 | 
> 39 | export class CodeBlockView implements NodeView {
  40 |   cm: CodeMirrorEditor
  41 |   dom?: HTMLElement
  42 |   updating: boolean = false

dispatchTransaction
src/components/ProsemirrorEditor/utils/buildEditor.ts:54

  51 |  * the Prosemiror state.
  52 |  * @param tr
  53 |  */
> 54 | dispatchTransaction(tr) {
     | ^  55 |   const nextState: any = view.state.apply(tr as any)
  56 | 
  57 |   view.updateState(nextState as any)

dispatch
node_modules/prosemirror-view/dist/index.es.js:4805
destroy
src/plugins/CodeBlock/nodeView.ts:39

  36 |   CodeMirror = require('codemirror')
  37 | }
  38 | 
> 39 | export class CodeBlockView implements NodeView {
  40 |   cm: CodeMirrorEditor
  41 |   dom?: HTMLElement
  42 |   updating: boolean = false

dispatchTransaction
src/components/ProsemirrorEditor/utils/buildEditor.ts:54

  51 |  * the Prosemiror state.
  52 |  * @param tr
  53 |  */
> 54 | dispatchTransaction(tr) {
     | ^  55 |   const nextState: any = view.state.apply(tr as any)
  56 | 
  57 |   view.updateState(nextState as any)

dispatch
node_modules/prosemirror-view/dist/index.es.js:4805
destroy
src/plugins/CodeBlock/nodeView.ts:39

  36 |   CodeMirror = require('codemirror')
  37 | }
  38 | 
> 39 | export class CodeBlockView implements NodeView {
  40 |   cm: CodeMirrorEditor
  41 |   dom?: HTMLElement
  42 |   updating: boolean = false

dispatchTransaction
src/components/ProsemirrorEditor/utils/buildEditor.ts:54

  51 |  * the Prosemiror state.
  52 |  * @param tr
  53 |  */
> 54 | dispatchTransaction(tr) {
     | ^  55 |   const nextState: any = view.state.apply(tr as any)
  56 | 
  57 |   view.updateState(nextState as any)

dispatch
node_modules/prosemirror-view/dist/index.es.js:4805
destroy
src/plugins/CodeBlock/nodeView.ts:39

  36 |   CodeMirror = require('codemirror')
  37 | }
  38 | 
> 39 | export class CodeBlockView implements NodeView {
  40 |   cm: CodeMirrorEditor
  41 |   dom?: HTMLElement
  42 |   updating: boolean = false

dispatchTransaction
src/components/ProsemirrorEditor/utils/buildEditor.ts:54

  51 |  * the Prosemiror state.
  52 |  * @param tr
  53 |  */
> 54 | dispatchTransaction(tr) {
     | ^  55 |   const nextState: any = view.state.apply(tr as any)
  56 | 
  57 |   view.updateState(nextState as any)

dispatch
node_modules/prosemirror-view/dist/index.es.js:4805
destroy
src/plugins/CodeBlock/nodeView.ts:39

  36 |   CodeMirror = require('codemirror')
  37 | }
  38 | 
> 39 | export class CodeBlockView implements NodeView {
  40 |   cm: CodeMirrorEditor
  41 |   dom?: HTMLElement
  42 |   updating: boolean = false

dispatchTransaction
src/components/ProsemirrorEditor/utils/buildEditor.ts:54

  51 |  * the Prosemiror state.
  52 |  * @param tr
  53 |  */
> 54 | dispatchTransaction(tr) {
     | ^  55 |   const nextState: any = view.state.apply(tr as any)
  56 | 
  57 |   view.updateState(nextState as any)

dispatch
node_modules/prosemirror-view/dist/index.es.js:4805
destroy
src/plugins/CodeBlock/nodeView.ts:39

  36 |   CodeMirror = require('codemirror')
  37 | }
  38 | 
> 39 | export class CodeBlockView implements NodeView {
  40 |   cm: CodeMirrorEditor
  41 |   dom?: HTMLElement
  42 |   updating: boolean = false

dispatchTransaction
src/components/ProsemirrorEditor/utils/buildEditor.ts:54

  51 |  * the Prosemiror state.
  52 |  * @param tr
  53 |  */
> 54 | dispatchTransaction(tr) {
     | ^  55 |   const nextState: any = view.state.apply(tr as any)
  56 | 
  57 |   view.updateState(nextState as any)

dispatch
node_modules/prosemirror-view/dist/index.es.js:4805
destroy
src/plugins/CodeBlock/nodeView.ts:39

  36 |   CodeMirror = require('codemirror')
  37 | }
  38 | 
> 39 | export class CodeBlockView implements NodeView {
  40 |   cm: CodeMirrorEditor
  41 |   dom?: HTMLElement
  42 |   updating: boolean = false

dispatchTransaction
src/components/ProsemirrorEditor/utils/buildEditor.ts:54

  51 |  * the Prosemiror state.
  52 |  * @param tr
  53 |  */
> 54 | dispatchTransaction(tr) {
     | ^  55 |   const nextState: any = view.state.apply(tr as any)
  56 | 
  57 |   view.updateState(nextState as any)

dispatch
node_modules/prosemirror-view/dist/index.es.js:4805
destroy
src/plugins/CodeBlock/nodeView.ts:39

  36 |   CodeMirror = require('codemirror')
  37 | }
  38 | 
> 39 | export class CodeBlockView implements NodeView {
  40 |   cm: CodeMirrorEditor
  41 |   dom?: HTMLElement
  42 |   updating: boolean = false

This screen is visible only in development. It will not appear if the app crashes in production.
Open your browser’s developer console to further inspect this error.  Click the 'X' or hit ESC to dismiss this message.
```

</details>

**Comment by jpuri** [comment]
Fixes https://github.com/tinacms/tinacms/issues/1276

**Comment by jpuri** [comment]
PR is updated with more fixes, I plan to do some more testing before moving it for review.

File: packages/demo-gatsby/content/blog/open-authoring-is-coming/index.md [hunk 1]
```diff
@@ -6,9 +6,4 @@ description: Edit websites in a single click!
 thumbnail: ../../assets/profile-pic.jpg
 blocks: []
 ---
-
-TinaCMS is capable of allowing visitors to [edit a website they're browsing](https://tinacms.org/blog/introducing-visual-open-authoring) and submit changes, thanks to a GitHub flow behind the scene.
-
-Contributing to an open-source documentation project has never been easier, TinaCMS will fork, save your changes and open a pull request for you. All you have to do is visually edit the content.
-
-![Visual authoring](https://tinacms.org/gif/open-auth.gif)
+**wecfsdfyyyyyyyyyyeeeeyyytttttttoooo111111uiiiii**
```
Comment (DirtyF) [inline packages/demo-gatsby/content/blog/open-authoring-is-coming/index.md:9]: We don't want to commit this :)

Comment (jpuri) [reply packages/demo-gatsby/content/blog/open-authoring-is-coming/index.md:9]: Yep I will fix it.

**Comment by jpuri** [comment]
Hey @DirtyF, @ncphillips : PR is ready for review and merge.

**Comment by DirtyF** [comment]
Tested on gatsby demo and run into an error when trying to escape the code block by typing SHIT+ENTER

```
TypeError: above.defaultContentType is not a function
exitCodeHard
src/plugins/CodeBlock/commands/index.ts:19

  16 | 
  17 | */
  18 | 
> 19 | import { EditorState, Selection, TextSelection } from 'prosemirror-state'
  20 | 
  21 | export function exitCodeHard(state: EditorState, dispatch: any) {
  22 |   const { $head } = state.selection

ShiftEnter
src/plugins/CodeBlock/nodeView.ts:39

  36 |   CodeMirror = require('codemirror')
  37 | }
  38 | 
> 39 | export class CodeBlockView implements NodeView {
  40 |   cm: CodeMirrorEditor
  41 |   dom?: HTMLElement
  42 |   updating: boolean = false
```

**Comment by DirtyF** [comment]
Weird can't reproduced it after refreshing the page.

One small thing I noticed is that you can't apply and unappy codeblock when you use the keyboard shortcut, it only works the first time, then you loose the selection.

https://www.loom.com/share/d2782f74e04d45539ccb1c3e4a7be685

**Comment by jpuri** [comment]
Good catch, that is actually only small issue that is still there, `alt-cmd-0` when used to remove code block results in focus going away from editor. I spent much time trying to fix this but fix for it was creating other issues, thus I would prefer to create another low prio ticket for this and merge this PR.

**Comment by ncphillips** [comment]
👍 

Some other stuff is taking priority right now but I'll try to get to this shortly

**Pull request merged**